agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,460 @@
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/raft"
+	"github.com/hashicorp/serf/serf"
+)
+
+// Delegate is the interface for the Autopilot mechanism
+type Delegate interface {
+	AutopilotConfig() *Config
+	FetchStats(context.Context, []serf.Member) map[string]*ServerStats
+	IsServer(serf.Member) (bool, *ServerInfo)
+	NotifyHealth(OperatorHealthReply)
+	PromoteNonVoters(*Config, OperatorHealthReply) ([]raft.Server, error)
+	Raft() *raft.Raft
+	Serf() *serf.Serf
+}
+
+// Autopilot is a mechanism for automatically managing the Raft
+// quorum using server health information along with updates from Serf gossip.
+// For more information, see https://www.consul.io/docs/guides/autopilot.html
+type Autopilot struct {
+	logger   *log.Logger
+	delegate Delegate
+
+	interval       time.Duration
+	healthInterval time.Duration
+
+	clusterHealth     OperatorHealthReply
+	clusterHealthLock sync.RWMutex
+
+	removeDeadCh chan struct{}
+	shutdownCh   chan struct{}
+	waitGroup    sync.WaitGroup
+}
+
+type ServerInfo struct {
+	Name   string
+	ID     string
+	Addr   net.Addr
+	Build  version.Version
+	Status serf.MemberStatus
+}
+
+func NewAutopilot(logger *log.Logger, delegate Delegate, interval, healthInterval time.Duration) *Autopilot {
+	return &Autopilot{
+		logger:         logger,
+		delegate:       delegate,
+		interval:       interval,
+		healthInterval: healthInterval,
+		removeDeadCh:   make(chan struct{}),
+	}
+}
+
+func (a *Autopilot) Start() {
+	a.shutdownCh = make(chan struct{})
+	a.waitGroup = sync.WaitGroup{}
+	a.waitGroup.Add(1)
+
+	go a.run()
+}
+
+func (a *Autopilot) Stop() {
+	close(a.shutdownCh)
+	a.waitGroup.Wait()
+}
+
+// run periodically looks for nonvoting servers to promote and dead servers to remove.
+func (a *Autopilot) run() {
+	defer a.waitGroup.Done()
+
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(a.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-ticker.C:
+			autopilotConfig := a.delegate.AutopilotConfig()
+			if autopilotConfig == nil {
+				continue
+			}
+
+			// Skip the non-voter promotions unless all servers support the new APIs
+			minRaftProtocol, err := a.MinRaftProtocol()
+			if err != nil {
+				a.logger.Printf("[ERR] autopilot: error getting server raft protocol versions: %s", err)
+				continue
+			}
+			if minRaftProtocol >= 3 {
+				promotions, err := a.delegate.PromoteNonVoters(autopilotConfig, a.GetClusterHealth())
+				if err != nil {
+					a.logger.Printf("[ERR] autopilot: Error checking for non-voters to promote: %s", err)
+				}
+				if err := a.handlePromotions(promotions); err != nil {
+					a.logger.Printf("[ERR] autopilot: Error handling promotions: %s", err)
+				}
+			}
+
+			if err := a.pruneDeadServers(); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		case <-a.removeDeadCh:
+			autopilotConfig := a.delegate.AutopilotConfig()
```

Comment:
You didn't add the `if` to gate the `pruneDeadServers()` call (here nor up in the previous clause). After seeing this, I'm thinking maybe it was better the old way where you pass the config and check it in there. Sorry for the flip-flop.

I was thinking you could do the same for the above clause to simplify it (because it's easy to miss a `continue`. You could make a `promoteServers()` helper that takes the config and check the Raft protocol in there, which would de-nest the `switch` and avoid lots of `continue` statements.